core/data/core.joke

```diff
@@ -4189,3 +4196,190 @@
   {:added "1.0"}
   []
   (joker-version__))
+
+(defn- check-valid-options
+  "Throws an exception if the given option map contains keys not listed
+  as valid, else returns nil."
+  [options & valid-keys]
+  (when (seq (apply disj (apply hash-set (keys options)) valid-keys))
+    (throw
+      (ex-info ; was: (IllegalArgumentException.
+        (apply str "Only these options are valid: "
+          (first valid-keys)
+          (map #(str ", " %) (rest valid-keys))) {}))))
+
+;;multimethods
+(defn- space-str
+  "Quick and dirty version of joker.string/join with a single space as the separator, because joker.core probably shouldn't depend on joker.string"
+  ([] "")
+  ([a] (str " " a))
+  ([a & more] (str " " a (apply space-str more))))
+
+(defn- multifn-nomatch
+  [name args]
+  (throw (ex-info (str "No method in multimethod for dispatch value: (" name (apply space-str args) ")") {})))
+
+(defn- new-multifn
+  [name dispatch-fn default hierarchy]
+  (when hierarchy
+    (throw (ex-info ":hierarchy not yet supported by joker.core/defmulti" {})))
+  (let [mfatom (atom {})]
+    (with-meta
+      (fn [& args]
+        (let [method (get @mfatom
+                          (apply dispatch-fn args)
+                          (get @mfatom
+                               default
+                               (fn [& args] (multifn-nomatch name args))))]
+          (apply method args)))
+      {:name name :dispatch-fn dispatch-fn :default default :ns *ns* :method-table mfatom})))
+
+(defn- multifn-swap-method-table-vals!
+  [multifn ^Callable f & args]
+  (let [mfm (meta multifn)
+        mfatom (:method-table mfm)]
+    (apply swap-vals! mfatom f args)))
+
+(defn- multifn-reset
```

Comment:
This and below private functions can and should be inlined in their public counterparts IMO.

Reply:
I'm okay with that, but note that I used Clojure's own defmulti code as a template for this effort, and kept the separation of public-function definitions from private-function implementations I saw there (and later saw that ClojureScript copied as well).

I don't mind combining them as suggested, but first want to make sure you're okay with moving the code base further afield from upstream.

(If defmulti was a rapidly moving target, i.e. a hotbed of innovation, I'd be more concerned about doing that!)

Reply:
The difference is that both Clojure and ClojureScript call methods of `IMultiFn` (or `MultiFn`), rather than private functions. This makes sense since the functionality is encapsulated in special types. Since Joker's implementation doesn't use special types, I think this addition layer is unnecessary.

Reply:
Yes, definitely not needed, and the code is easier to read that way.

Done!